File: b/microfundamentos/AOP/2-Ordenacao/heap/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define TAM 9

typedef struct {
    int array[TAM];
    int tam;
} Heap;

Heap* criarHeap() {
    Heap* h = (Heap*)malloc(sizeof(Heap));
    h->tam = 0;
    
    return h;
}

void mostrarHeap(Heap* h) {
    if(h->tam == 0) {
        printf("Heap vazia\n\n"); return;
    }

    printf("Heap tam: %d -> ", h->tam);
    for(int i = 0; i < h->tam; i++) printf("%d ", h->array[i]); printf("\n");
}

void ajustarHeapMax(Heap* h) {

    int indexFi = h->tam-1;
    int indexPai = (indexFi-1)/2;

    while(indexFi > 0) {

        if(h->array[indexFi] > h->array[indexPai]) {
            int aux = h->array[indexFi]; h->array[indexFi] = h->array[indexPai]; h->array[indexPai] = aux;
        }

        indexFi = indexPai;
        indexPai = (indexFi-1)/2;

    }
    
}

void ajustarHeapMin(Heap* h) {

    int indexFi = h->tam-1;
    int indexPai = (indexFi-1)/2;

    while(indexFi > 0) {

        if(h->array[indexFi] < h->array[indexPai]) {
            int aux = h->array[indexFi]; h->array[indexFi] = h->array[indexPai]; h->array[indexPai] = aux;
        }

        indexFi = indexPai;
        indexPai = (indexFi-1)/2;

    }
}

void inserirHeap(Heap* h, int elem) {
    
    if(h->tam >= TAM) {
        printf("Heap cheia\n"); return;
    }
    
    h->array[h->tam++] = elem;

    mostrarHeap(h);
    ajustarHeapMax(h);
    //ajustarHeapMin(h);
    mostrarHeap(h);
}

void inserirCresHeap(Heap* h) {

    if(h->tam >= TAM) {
        printf("Heap cheia\n"); return;
    }

    for(int i = 0; h->tam < TAM; i++) {
        h->array[h->tam++] = i;
    }

    mostrarHeap(h);
    ajustarHeapMax(h);
    //ajustarHeapMin(h);
    mostrarHeap(h);
}

void inserirDecresHeap(Heap* h) {

    if(h->tam >= TAM) {
        printf("Heap cheia\n"); return;
    }

    for(int i = TAM; h->tam < TAM; i--) {
        h->array[h->tam++] = i;
    }

    mostrarHeap(h);
    //ajustarHeapMax(h);
    ajustarHeapMin(h);
    mostrarHeap(h);
}


int main(void) {
    
    Heap* minhaheap = criarHeap();
    
    mostrarHeap(minhaheap);

    //inserirCresHeap(minhaheap);
    inserirDecresHeap(minhaheap);

    mostrarHeap(minhaheap);

    
    return 0;
}

File: b/microfundamentos/AOP/2-Ordenacao/heap/Main.java
import java.io.IOException;
import java.util.Scanner;

class Heap {
    private static final int TAM = 9;
    int[] array;
    int tam;

    public Heap() {
        this.tam = 0;
        this.array = new int[TAM];
    }

    public int getTAM() {
        return TAM;
    }

    public void mostrarHeap() {
        if(tam == 0) {
            System.out.println("Heap vazia\n"); return;
        }

        System.out.print("Heap tam: "+ tam + " -> ");
        for(int i = 0; i < tam; i++) System.out.print(array[i] + " "); System.out.println();
    }

    public void ajustarHeapMax() {

        int indexFi = tam-1;
        int indexPai = (indexFi-1)/2;

        while(indexFi > 0) {

            if(array[indexFi] > array[indexPai]) {
                int aux = array[indexFi]; array[indexFi] = array[indexPai]; array[indexPai] = aux;
            }

            indexFi = indexPai;
            indexPai = (indexFi-1)/2;
        }
    }

    public void ajustarHeapMin() {

        int indexFi = tam-1;
        int indexPai = (indexFi-1)/2;

        while(indexFi > 0) {

            if(array[indexFi] < array[indexPai]) {
                int aux = array[indexFi]; array[indexFi] = array[indexPai]; array[indexPai] = aux;
            }

            indexFi = indexPai;
            indexPai = (indexFi-1)/2;
        }
    }

    public void inserirHeap(int elem) {
        if(tam >= TAM) {
            System.out.println("Heap cheia"); return;
        }

        array[tam++] = elem;

        this.mostrarHeap();
        //this.ajustarHeapMax();
        this.ajustarHeapMin();
        this.mostrarHeap();
    }

    public void inserirCrescHeap() {
        if(tam >= TAM) {
            System.out.println("Heap cheia"); return;
        }

        for(int i = 0; tam < TAM; i++) {
            array[tam++] = i;
        }

        this.mostrarHeap();
        this.ajustarHeapMax();
        //this.ajustarHeapMin();
        this.mostrarHeap();
    }

    public void inserirDerescHeap() {
        if(tam >= TAM) {
            System.out.println("Heap cheia"); return;
        }

        for(int i = TAM; tam < TAM; i--) {
            array[tam++] = i;
        }

        this.mostrarHeap();
        //this.ajustarHeapMax();
        this.ajustarHeapMin();
        this.mostrarHeap();
    }
}

class Main {

    public static void main(String[] args) {

        Heap minhaHeap = new Heap();

        minhaHeap.mostrarHeap();

        minhaHeap.inserirCrescHeap();
        //minhaHeap.inserirDerescHeap();

        minhaHeap.mostrarHeap();

        System.out.println(minhaHeap.getTAM());
    }
}